File: concepts/08-security/node/jwt-auth.js
/**
 * JWT Authentication Implementation
 * Demonstrates JWT token generation, validation, and refresh
 */

const crypto = require('crypto');

/**
 * JWT Token Manager
 */
class JWTAuth {
  constructor(secret, options = {}) {
    this.secret = secret || this.generateSecret();
    this.algorithm = options.algorithm || 'HS256';
    this.accessTokenExpiry = options.accessTokenExpiry || 3600; // 1 hour
    this.refreshTokenExpiry = options.refreshTokenExpiry || 604800; // 7 days
    this.refreshTokens = new Map(); // Store refresh tokens
  }

  /**
   * Generate secret key
   */
  generateSecret() {
    return crypto.randomBytes(32).toString('hex');
  }

  /**
   * Base64 URL encode
   */
  base64UrlEncode(str) {
    return Buffer.from(str)
      .toString('base64')
      .replace(/\+/g, '-')
      .replace(/\//g, '_')
      .replace(/=/g, '');
  }

  /**
   * Base64 URL decode
   */
  base64UrlDecode(str) {
    str = str.replace(/-/g, '+').replace(/_/g, '/');
    while (str.length % 4) {
      str += '=';
    }
    return Buffer.from(str, 'base64').toString('utf8');
  }

  /**
   * Create HMAC signature
   */
  createSignature(header, payload) {
    const data = `${header}.${payload}`;
    const signature = crypto
      .createHmac('sha256', this.secret)
      .update(data)
      .digest('base64')
      .replace(/\+/g, '-')
      .replace(/\//g, '_')
      .replace(/=/g, '');
    return signature;
  }

  /**
   * Verify signature
   */
  verifySignature(header, payload, signature) {
    const expectedSignature = this.createSignature(header, payload);
    return crypto.timingSafeEqual(
      Buffer.from(signature),
      Buffer.from(expectedSignature)
    );
  }

  /**
   * Generate access token
   */
  generateAccessToken(payload) {
    const header = {
      alg: this.algorithm,
      typ: 'JWT'
    };

    const now = Math.floor(Date.now() / 1000);
    const tokenPayload = {
      ...payload,
      iat: now,
      exp: now + this.accessTokenExpiry,
      type: 'access'
    };

    const encodedHeader = this.base64UrlEncode(JSON.stringify(header));
    const encodedPayload = this.base64UrlEncode(JSON.stringify(tokenPayload));
    const signature = this.createSignature(encodedHeader, encodedPayload);

    return `${encodedHeader}.${encodedPayload}.${signature}`;
  }

  /**
   * Generate refresh token
   */
  generateRefreshToken(userId) {
    const token = crypto.randomBytes(32).toString('hex');
    const expiresAt = Date.now() + (this.refreshTokenExpiry * 1000);
    
    this.refreshTokens.set(token, {
      userId,
      expiresAt,
      createdAt: Date.now()
    });

    return token;
  }

  /**
   * Verify and decode token
   */
  verifyToken(token) {
    try {
      const parts = token.split('.');
      if (parts.length !== 3) {
        throw new Error('Invalid token format');
      }

      const [encodedHeader, encodedPayload, signature] = parts;

      // Verify signature
      if (!this.verifySignature(encodedHeader, encodedPayload, signature)) {
        throw new Error('Invalid signature');
      }

      // Decode payload
      const payload = JSON.parse(this.base64UrlDecode(encodedPayload));

      // Check expiration
      if (payload.exp && payload.exp < Math.floor(Date.now() / 1000)) {
        throw new Error('Token expired');
      }

      return payload;
    } catch (error) {
      throw new Error(`Token verification failed: ${error.message}`);
    }
  }

  /**
   * Refresh access token using refresh token
   */
  refreshAccessToken(refreshToken) {
    const tokenData = this.refreshTokens.get(refreshToken);
    
    if (!tokenData) {
      throw new Error('Invalid refresh token');
    }

    if (Date.now() > tokenData.expiresAt) {
      this.refreshTokens.delete(refreshToken);
      throw new Error('Refresh token expired');
    }

    // Generate new access token
    const newAccessToken = this.generateAccessToken({
      userId: tokenData.userId
    });

    return newAccessToken;
  }

  /**
   * Revoke refresh token
   */
  revokeRefreshToken(refreshToken) {
    return this.refreshTokens.delete(refreshToken);
  }

  /**
   * Cleanup expired refresh tokens
   */
  cleanupExpiredTokens() {
    const now = Date.now();
    for (const [token, data] of this.refreshTokens.entries()) {
      if (now > data.expiresAt) {
        this.refreshTokens.delete(token);
      }
    }
  }

  /**
   * Authenticate user and generate tokens
   */
  authenticate(userId, userData = {}) {
    const accessToken = this.generateAccessToken({
      userId,
      ...userData
    });

    const refreshToken = this.generateRefreshToken(userId);

    return {
      accessToken,
      refreshToken,
      tokenType: 'Bearer',
      expiresIn: this.accessTokenExpiry
    };
  }
}

// Example usage
function demonstrateJWTAuth() {
  console.log('=== JWT Authentication ===\n');

  const jwtAuth = new JWTAuth('my-secret-key', {
    accessTokenExpiry: 3600, // 1 hour
    refreshTokenExpiry: 604800 // 7 days
  });

  // Authenticate user
  console.log('1. Authenticating user...');
  const tokens = jwtAuth.authenticate('user123', { username: 'john', role: 'admin' });
  console.log('Tokens generated:', {
    accessToken: tokens.accessToken.substring(0, 50) + '...',
    refreshToken: tokens.refreshToken.substring(0, 20) + '...',
    expiresIn: tokens.expiresIn
  });

  // Verify access token
  console.log('\n2. Verifying access token...');
  try {
    const payload = jwtAuth.verifyToken(tokens.accessToken);
    console.log('Token verified:', payload);
  } catch (error) {
    console.error('Verification failed:', error.message);
  }

  // Refresh access token
  console.log('\n3. Refreshing access token...');
  try {
    const newAccessToken = jwtAuth.refreshAccessToken(tokens.refreshToken);
    console.log('New access token generated:', newAccessToken.substring(0, 50) + '...');
    
    const newPayload = jwtAuth.verifyToken(newAccessToken);
    console.log('New token payload:', newPayload);
  } catch (error) {
    console.error('Refresh failed:', error.message);
  }

  // Test expired token
  console.log('\n4. Testing token expiration...');
  const expiredToken = jwtAuth.generateAccessToken({ userId: 'user123' });
  // Simulate expiration by modifying expiry
  const parts = expiredToken.split('.');
  const payload = JSON.parse(Buffer.from(parts[1], 'base64').toString());
  payload.exp = Math.floor(Date.now() / 1000) - 100; // Expired 100 seconds ago
  parts[1] = Buffer.from(JSON.stringify(payload)).toString('base64');
  const expiredTokenModified = parts.join('.');
  
  try {
    jwtAuth.verifyToken(expiredTokenModified);
  } catch (error) {
    console.log('Correctly detected expired token:', error.message);
  }
}

if (require.main === module) {
  demonstrateJWTAuth();
}

module.exports = JWTAuth;

File: concepts/08-security/python/jwt_auth.py
"""
JWT Authentication Implementation
Demonstrates JWT token generation, validation, and refresh
"""
import hmac
import hashlib
import base64
import json
import secrets
import time
from typing import Dict, Any, Optional
from dataclasses import dataclass


@dataclass
class RefreshTokenData:
    """Refresh token data"""
    user_id: str
    expires_at: float
    created_at: float


class JWTAuth:
    """JWT Authentication implementation"""
    
    def __init__(self, secret: Optional[str] = None, options: Dict[str, Any] = None):
        self.secret = secret or self._generate_secret()
        self.algorithm = (options or {}).get('algorithm', 'HS256')
        self.access_token_expiry = (options or {}).get('access_token_expiry', 3600)  # 1 hour
        self.refresh_token_expiry = (options or {}).get('refresh_token_expiry', 604800)  # 7 days
        self.refresh_tokens: Dict[str, RefreshTokenData] = {}
    
    def _generate_secret(self) -> str:
        """Generate secret key"""
        return secrets.token_hex(32)
    
    def _base64_url_encode(self, data: str) -> str:
        """Base64 URL encode"""
        return base64.urlsafe_b64encode(
            data.encode('utf-8')
        ).decode('utf-8').rstrip('=')
    
    def _base64_url_decode(self, data: str) -> str:
        """Base64 URL decode"""
        padding = 4 - len(data) % 4
        if padding != 4:
            data += '=' * padding
        return base64.urlsafe_b64decode(data).decode('utf-8')
    
    def _create_signature(self, header: str, payload: str) -> str:
        """Create HMAC signature"""
        data = f'{header}.{payload}'
        signature = hmac.new(
            self.secret.encode('utf-8'),
            data.encode('utf-8'),
            hashlib.sha256
        ).digest()
        return self._base64_url_encode(base64.b64encode(signature).decode('utf-8'))
    
    def _verify_signature(self, header: str, payload: str, signature: str) -> bool:
        """Verify signature"""
        expected_signature = self._create_signature(header, payload)
        return hmac.compare_digest(signature, expected_signature)
    
    def generate_access_token(self, payload: Dict[str, Any]) -> str:
        """Generate access token"""
        header = {
            'alg': self.algorithm,
            'typ': 'JWT'
        }
        
        now = int(time.time())
        token_payload = {
            **payload,
            'iat': now,
            'exp': now + self.access_token_expiry,
            'type': 'access'
        }
        
        encoded_header = self._base64_url_encode(json.dumps(header))
        encoded_payload = self._base64_url_encode(json.dumps(token_payload))
        signature = self._create_signature(encoded_header, encoded_payload)
        
        return f'{encoded_header}.{encoded_payload}.{signature}'
    
    def generate_refresh_token(self, user_id: str) -> str:
        """Generate refresh token"""
        token = secrets.token_hex(32)
        expires_at = time.time() + self.refresh_token_expiry
        
        self.refresh_tokens[token] = RefreshTokenData(
            user_id=user_id,
            expires_at=expires_at,
            created_at=time.time()
        )
        
        return token
    
    def verify_token(self, token: str) -> Dict[str, Any]:
        """Verify and decode token"""
        try:
            parts = token.split('.')
            if len(parts) != 3:
                raise ValueError('Invalid token format')
            
            encoded_header, encoded_payload, signature = parts
            
            # Verify signature
            if not self._verify_signature(encoded_header, encoded_payload, signature):
                raise ValueError('Invalid signature')
            
            # Decode payload
            payload = json.loads(self._base64_url_decode(encoded_payload))
            
            # Check expiration
            if payload.get('exp') and payload['exp'] < time.time():
                raise ValueError('Token expired')
            
            return payload
        except Exception as e:
            raise ValueError(f'Token verification failed: {e}')
    
    def refresh_access_token(self, refresh_token: str) -> str:
        """Refresh access token using refresh token"""
        token_data = self.refresh_tokens.get(refresh_token)
        
        if not token_data:
            raise ValueError('Invalid refresh token')
        
        if time.time() > token_data.expires_at:
            del self.refresh_tokens[refresh_token]
            raise ValueError('Refresh token expired')
        
        # Generate new access token
        new_access_token = self.generate_access_token({
            'userId': token_data.user_id
        })
        
        return new_access_token
    
    def revoke_refresh_token(self, refresh_token: str) -> bool:
        """Revoke refresh token"""
        if refresh_token in self.refresh_tokens:
            del self.refresh_tokens[refresh_token]
            return True
        return False
    
    def cleanup_expired_tokens(self):
        """Cleanup expired refresh tokens"""
        now = time.time()
        expired = [
            token for token, data in self.refresh_tokens.items()
            if now > data.expires_at
        ]
        for token in expired:
            del self.refresh_tokens[token]
    
    def authenticate(self, user_id: str, user_data: Dict[str, Any] = None) -> Dict[str, Any]:
        """Authenticate user and generate tokens"""
        access_token = self.generate_access_token({
            'userId': user_id,
            **(user_data or {})
        })
        
        refresh_token = self.generate_refresh_token(user_id)
        
        return {
            'accessToken': access_token,
            'refreshToken': refresh_token,
            'tokenType': 'Bearer',
            'expiresIn': self.access_token_expiry
        }


def demonstrate_jwt_auth():
    """Demonstrate JWT authentication"""
    print('=== JWT Authentication ===\n')
    
    jwt_auth = JWTAuth('my-secret-key', {
        'access_token_expiry': 3600,  # 1 hour
        'refresh_token_expiry': 604800  # 7 days
    })
    
    # Authenticate user
    print('1. Authenticating user...')
    tokens = jwt_auth.authenticate('user123', {'username': 'john', 'role': 'admin'})
    print('Tokens generated:', {
        'accessToken': tokens['accessToken'][:50] + '...',
        'refreshToken': tokens['refreshToken'][:20] + '...',
        'expiresIn': tokens['expiresIn']
    })
    
    # Verify access token
    print('\n2. Verifying access token...')
    try:
        payload = jwt_auth.verify_token(tokens['accessToken'])
        print('Token verified:', payload)
    except ValueError as e:
        print(f'Verification failed: {e}')
    
    # Refresh access token
    print('\n3. Refreshing access token...')
    try:
        new_access_token = jwt_auth.refresh_access_token(tokens['refreshToken'])
        print('New access token generated:', new_access_token[:50] + '...')
        
        new_payload = jwt_auth.verify_token(new_access_token)
        print('New token payload:', new_payload)
    except ValueError as e:
        print(f'Refresh failed: {e}')
    
    # Test invalid token
    print('\n4. Testing invalid token...')
    try:
        jwt_auth.verify_token('invalid.token.here')
    except ValueError as e:
        print('Correctly detected invalid token:', e)


if __name__ == '__main__':
    demonstrate_jwt_auth()

